packages/daemon/src/types.d.ts

```diff
@@ -718,6 +717,37 @@ type FormulateNumberedHostParams = {
   networksDirectoryId: string;
 };
 
+export type Provide = <T extends string>(
+  id: string,
+  type?: T,
+) => Promise<
+  T extends 'endo'
+    ? EndoBootstrap
+    : T extends 'pet-store'
+    ? PetStore
+    : T extends 'peer'
+    ? EndoGateway
+    : T extends 'directory'
+    ? EndoDirectory
+    : T extends 'network'
+    ? EndoNetwork
+    : T extends 'host'
+    ? EndoHost
+    : T extends 'hub'
+    ? NameHub
+    : T extends 'readable-blob'
+    ? EndoReadable
+    : T extends 'worker'
+    ? EndoWorker
+    : T extends 'agent'
+    ? EndoAgent
+    : T extends 'guest'
+    ? EndoGuest
+    : T extends 'invitation'
+    ? Invitation
+    : unknown
+>;
```

Comment:
A couple of things regarding this one. First, I propose we reimplement it as follows:

1. Establish a private type that only exists to map `Formula['type']` to the type of the value instantiated by a formula:

```ts
type FormulaForValue = {
  endo: EndoBootstrap,
  host: EndoHost,
  guest: EndoGuest,
  worker: EndoWorker,
  // ...
};
```

2. Replace the ternary chain above with:
```ts
export type Provide = <T extends Formula['type'], U extends FormulaForValue[T]>(
  id: string,
  type?: T,
) => Promise<U>;
```

This has the benefit of being more succinct and partially self-validating since the definition of `Provide` errors if `FormulaForValue` does not have a key for every value in the string literal union `Formula['type']`. (It's still possible for `keyof FormulaForValue` to superset `Formula['type']`, but that will cause type errors at the point of use.)

Second, as it happens, `Formula['type']` in fact **does not** include all of the string literals used as the second argument to `provide()`, specifically the following:
- `agent`
- `hub`
- `network`

I had thought that there was a bijection between "formula types" and "values that can be provided". I think we should attempt to eliminate this discrepancy if possible. If we don't want to do that now, we can implement my above suggestion with the following band-aid:
```ts
type Providables = Formula['type'] & 'agent' | 'hub' | 'network';
// FormulaForValue modified as necessary
export type Provide = <T extends Providables, U extends FormulaForValue[T]>( /* ... */
```